Testnet/Airdrop/Airdrop/Airdrop.cs

```diff
@@ -0,0 +1,202 @@
+    /// </summary>
+    /// <param name="smartContractState">The execution state for the contract.</param>
+    /// <param name="tokenContractAddress">The smart contract address of the token being airdropped.</param>
+    /// <param name="totalSupply">The total amount that will be airdropped, amount will be divided amongst registrants.</param>
+    /// <param name="endBlock">The block that ends the sign up period and allows withdrawing, can use 0 to manually end airdrop using <see cref="CloseRegistration"/>.</param>
+    public Airdrop(
+        ISmartContractState smartContractState,
+        Address tokenContractAddress,
+        ulong totalSupply,
+        ulong endBlock
+    ) : base(smartContractState)
+    {
+        TotalSupply = totalSupply;
+        TokenContractAddress = tokenContractAddress;
+        EndBlock = endBlock;
+        Owner = Message.Sender;
+    }
+
+    /// <summary>The contract address of the token that will be distributed. This smart contracts address must hold the TotalSupply at this address.</summary>
+    public Address TokenContractAddress
+    {
+        get => PersistentState.GetAddress(nameof(TokenContractAddress));
+        private set => PersistentState.SetAddress(nameof(TokenContractAddress), value);
+    }
+
+    /// <summary>The total supply of the token that will be distributed during this airdrop.</summary>
+    public ulong TotalSupply
+    {
+        get => PersistentState.GetUInt64(nameof(TotalSupply));
+        private set => PersistentState.SetUInt64(nameof(TotalSupply), value);
+    }
+
+    /// <summary>
+    /// Used to automatically close registration based on a specified block number. If the EndBlock is 0, the
+    /// registration period for the airdrop must be manually ended by the owner using <see cref="CloseRegistration"/>
+    /// </summary>
+    public ulong EndBlock
+    {
+        get => PersistentState.GetUInt64(nameof(EndBlock));
+        private set => PersistentState.SetUInt64(nameof(EndBlock), value);
+    }
+
+    /// <summary>Address of the owner of this contract, used to authenticate some methods.</summary>
+    public Address Owner
+    {
+        get => PersistentState.GetAddress(nameof(Owner));
+        private set => PersistentState.SetAddress(nameof(Owner), value);
+    }
+
+    /// <summary>The total number of registrants for this airdrop.</summary>
+    public ulong NumberOfRegistrants
+    {
+        get => PersistentState.GetUInt64(nameof(NumberOfRegistrants));
+        private set => PersistentState.SetUInt64(nameof(NumberOfRegistrants), value);
+    }
+
+    /// <summary>Calculates and sets the amount to distribute to each registrant.</summary>
+    public ulong AmountToDistribute
+    {
+        get
+        {
+            ulong amount = PersistentState.GetUInt64(nameof(AmountToDistribute));
+            if (amount == 0 && RegistrationIsClosed)
+            {
+                amount = TotalSupply / NumberOfRegistrants;
```

Comment:
This will lose tokens due to rounding. Consider: `TotalSupply = 11` and `NumberOfRegistrants = 5`.

Reply:
To elaborate: the correct amount will still be distributed to the participants, but there will be some fraction of the total supply left over if the amounts are not evenly divisible. This might be OK if there are alternate means for the owner to withdraw those tokens from the contract.